### ISSUE TITLE
Unable to read parquet data without _metadata

### ISSUE THREAD

**Issue by kwmsmith** [dask/fastparquet#137]
I'm unable to access parquet data (on s3, if that's relevant) that does not have `_metadata`.  This dataset was generated via spark 2.1 (databricks); according to their docs, a `_metadata` file is optional.  

    https://forums.databricks.com/questions/6494/do-you-need-a-metadata-file-in-a-parquet-folder-to.html

Versions:

```
$ conda list | grep 'dask\|fastparquet'
dask                      0.14.1                   py35_0    conda-forge
fastparquet               0.0.5                    py35_1    conda-forge
```

Traceback:

```
In [18]: dd.read_parquet("s3://aaa/bbb")
---------------------------------------------------------------------------
ClientError                               Traceback (most recent call last)
[...]/lib/python3.5/site-packages/s3fs/core.py in info(self, path, refresh)
    358                 out = self.s3.head_object(Bucket=bucket, Key=key,
--> 359                                           **self.req_kw)
    360                 out = {'ETag': out['ETag'], 'Key': '/'.join([bucket, key]),

[...]/lib/python3.5/site-packages/botocore/client.py in _api_call(self, *args, **kwargs)
    252             # The "self" in this scope is referring to the BaseClient.
--> 253             return self._make_api_call(operation_name, kwargs)
    254

[...]/lib/python3.5/site-packages/botocore/client.py in _make_api_call(self, operation_name, api_params)
    543             error_class = self.exceptions.from_code(error_code)
--> 544             raise error_class(parsed_response, operation_name)
    545         else:

ClientError: An error occurred (404) when calling the HeadObject operation: Not Found

During handling of the above exception, another exception occurred:

FileNotFoundError                         Traceback (most recent call last)
[...]/lib/python3.5/site-packages/fastparquet/api.py in __init__(self, fn, verify, open_with, sep)
     46             self.fn = fn2
---> 47             with open_with(fn2, 'rb') as f:
     48                 self._parse_header(f, verify)

[...]/lib/python3.5/site-packages/dask/bytes/core.py in __enter__(self)
    313         mode = self.mode.replace('t', '').replace('b', '') + 'b'
--> 314         f = f2 = self.myopen(self.path, mode=mode)
    315         CompressFile = merge(seekable_files, compress_files)[self.compression]

[...]/lib/python3.5/site-packages/dask/bytes/s3.py in open(self, path, mode, **kwargs)
     41         s3_path = bucket + path
---> 42         return S3FileSystem.open(self, s3_path, mode=mode)
     43

[...]/lib/python3.5/site-packages/s3fs/core.py in open(self, path, mode, block_size, acl, fill_cache)
    257         return S3File(self, path, mode, block_size=block_size, acl=acl,
--> 258                       fill_cache=fill_cache)
    259

[...]/lib/python3.5/site-packages/s3fs/core.py in __init__(self, s3, path, mode, block_size, acl, fill_cache)
    854             try:
--> 855                 self.size = self.info()['Size']
    856             except (ClientError, ParamValidationError):

[...]/lib/python3.5/site-packages/s3fs/core.py in info(self)
    860         """ File information about this path """
--> 861         return self.s3.info(self.path)
    862

[...]/lib/python3.5/site-packages/s3fs/core.py in info(self, path, refresh)
    364             except (ClientError, ParamValidationError):
--> 365                 raise FileNotFoundError(path)
    366

FileNotFoundError: aaa/bbb/_metadata

```

**Comment by martindurant** [comment]
Please try with `dd.read_parquet("s3://aaa/bbb/*.parquet")`. The reasoning here, is that we don't want to walk the whole directory structure and try to identify files that might be parquet, and load them to see if they are - we leave this choice of files to probe up to the user.

**Comment by kwmsmith** [comment]
OK, so I created a new env with dask master and fastparquet master, and used the `.../*.parquet` globstring.  It looks like that resolved my issue, thanks!

Do you have an ETA on the next release of fastparquet (and dask?).

For this particular dataset (about 200 parquet partitions, about 2 dozen columns, all nullable, approx 1e7 rows total) the `dd.read_parquet()` call takes about 3 minutes to complete.  I would have hoped that the `read_parquet()` call would be instantaneous -- any pointers on what's going on?  I'm happy to create a separate issue out of this if helpful.

**Comment by martindurant** [comment]
You have put your finger exactly on the problem: when there is no metadata file, every single component's metadata must be accessed in order just to know what range of values they have (and, in fact, if they are parquet files); on s3, this means many HTTP calls. It is possible that this could be skipped, and uniform schemas assumed, etc.
You could choose to use fastparquet.writer.merge to create a metadata file once, and thereafter access that, so you pay the cost up front just the once.

**Comment by martindurant** [comment]
btw: ~50,000 rows per row-group is pretty small. For example, fastparquet.write assumes 50,000,000 unless the user specifies otherwise; of course, this depends on any pre-filtering that might be wanted on reading and the number of columns.

**Comment by kwmsmith** [comment]
On your advice, I wrote out `_metadata` and `_common_metadata` files via `fastparquet.writer.merge`:

```
In [68]: fp.writer.merge(s3files[1:], verify_schema=False, open_with=s3.open)
```

Now when I `dd.read_parquet('s3://aaa/bbb')`, it returns almost instantaneously, so that definitely helped.

But doing a trivial column operation still takes a really long time (order of minutes for the smallish dataset mentioned previously):

```
df.col.value_counts().compute()
```

I'd expect this to be very fast, especially now that all the metadata is in place.  Thoughts?

**Comment by martindurant** [comment]
There is no way to do value_counts without opening at least one connection for each file in the dataset, even if the number of bytes that need to be downloaded from each is relatively small for one column of interest. It is setting up connections that takes the time, even if you have four or eight worker threads sending requests in parallel. This is why parquet shards tend to be somewhat large (say 64MB).
Note that there is very probably still some optimization possible in this regard within both dask and s3fs, the back-end connection library.

**Comment by mrocklin** [comment]
@kwmsmith if you're on a single machine you might want to try profiling with the single-threaded scheduler

```python
foo.compute(get=dask.async.get_sync)
```

This should allow you to get normal statistics on what's taking up time.  If, as @martindurant suggests, you're being dragged down by opening connections (this would not surprise me, Martin has a decent intuition for these systems) the profile information might provide sufficient feedback to inform optimization on your end.

**Comment by abedshantti** [comment]
I am getting this error for some reason. Was working fine a couple of months ago. I am only using dask.dataframe and I am trying to read all the parquet files in subfolders in my S3 bucket. 

```
df = dd.read_parquet("s3:aa/bb/folder=199/*.parquet")
```

However, I receive the following error:
```
FileNotFoundError: aa/bb/folder=199/part.0.parquet/_metadata
PermissionError: Access Denied
```

This is really strange because read works for the first few folders in my S3 directory, but fails with the ones that require more navigation to find. I can confirm that the folders and the parquet files in them, do exist in my S3 bucket. However, dask seems to require the metadata files, which do not exist,

**Comment by abedshantti** [comment]
Just to be clear, my directory has over 100,000 folders in my S3 bucket. Each of these folders has several parquet files. Dask manages to read the parquet files in the earlier folders (the ones that I can easily find by navigating through the pages in my S3 bucket directory (perhaps only the first 5,000 folders), but is unable to find the ones that I would struggle to find by simply navigating manually through the pages. I can't estimate the actual number of the ones that dask can properly read, since the folder names don't start from 0.
